Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,472 @@
+				resetIfInputIsTexture,
+				nullptr,
+				setTextureHandler);
+		}
+	}
+
+	auto strengthInput = getInputSocket(sourceNode, normalStrengthIdentifier);
+	if (strengthInput && setColorHandler) {
+		auto defaultValue = (bNodeSocketValueFloat*)strengthInput->default_value;
+		auto val = defaultValue->value;
+		setColorHandler(mu::float4{ val,val,val, val });
+	}
+}
+
+void msblenMaterialsExportHelper::handleDisplacementNode(const Material* mat,
+	ms::TextureType textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode)
+{
+	auto heightInput = getInputSocket(sourceNode, heightIdentifier);
+	if (heightInput && setTextureHandler)
+	{
+		setValueFromSocket(mat,
+			heightInput, textureType,
+			resetIfInputIsTexture,
+			setColorHandler,
+			setTextureHandler);
+	}
+
+	auto scaleInput = getInputSocket(sourceNode, scaleIdentifier);
+	if (scaleInput && setColorHandler)
+	{
+		auto defaultValue = (bNodeSocketValueFloat*)scaleInput->default_value;
+		auto val = defaultValue->value;
+		setColorHandler(mu::float4{ val,val,val, val });
+	}
+}
+
+void msblenMaterialsExportHelper::handlePassthrough(const Material* mat,
+	ms::TextureType textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode)
+{
+	auto imageInput = getInputSocket(sourceNode, colorIdentifier);
+	if (imageInput)
+	{
+		setValueFromSocket(mat,
+			imageInput, textureType,
+			resetIfInputIsTexture,
+			setColorHandler,
+			setTextureHandler);
+	}
+}
+
+void msblenMaterialsExportHelper::handleSocketValue(bNodeSocket* socket,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler)
+{
+	// The socket is using a direct value, not the output of another node.
+	// Clear any texture that was set:
+	if (setTextureHandler) {
+		setTextureHandler(ms::InvalidID);
+	}
+
+	if (setColorHandler) {
+		mu::float4 colorValue;
+
+		switch (socket->type)
+		{
+		case SOCK_RGBA: {
+			auto defaultValue = (bNodeSocketValueRGBA*)socket->default_value;
+			auto val = defaultValue->value;
+			colorValue = mu::float4{ val[0], val[1], val[2], val[3] };
+			break;
+		}
+		case SOCK_FLOAT: {
+			auto defaultValue = (bNodeSocketValueFloat*)socket->default_value;
+			auto val = defaultValue->value;
+			colorValue = mu::float4{ val, val, val, val };
+			break;
+		}
+		}
+		setColorHandler(colorValue);
+	}
+}
+
+void msblenMaterialsExportHelper::setValueFromSocket(const Material* mat,
+	bNodeSocket* socket,
+	ms::TextureType textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler)
+{
+	// If there is an image linked to the socket, send that as a texture:
+	if (socket->link) {
```

Comment:
Can you check the negation and return after handling the socket value? Should help reducing nesting

Reply:
Done